File: 45. Jump Game II/Jump.py
'''
https://leetcode.com/problems/jump-game-ii/
'''


from typing import List


class Solution:
    # O(n*m) time, m--> the biggest jump in the nums
    # O(1) space,
    # Approach: DP, top bottom,
    def jump(self, nums: List[int]) -> int:
        n = len(nums)
        i = 0
        jumps = 0
        
        while i < n-1:
            jump = nums[i]
            jumps +=1
            if i + nums[i] > n-2:
                break
            max_jump_index = i+1
            for j in range(i+2, i+jump+1):
                if nums[j] == 0:    continue
                if j + nums[j] > max_jump_index + nums[max_jump_index]:
                    # print(j)
                    max_jump_index = j
                    
            i = max_jump_index
            
        return jumps

File: 45. Jump Game II/jump.rs
use std::cmp;


impl Solution {

    // O(n*max(nums[i])) time,
    // O(n) space,
    // Approach: dp, bottom up, 
    pub fn jump(nums: Vec<i32>) -> i32 {
        if nums.len() < 2 {
            return 0;
        }
        
        let mut dp = vec![i32::MAX-2;nums.len()];
        dp[nums.len()-1] = 0;
        dp[nums.len()-2] = 1;
        
        
        for i in (0..nums.len()-1).rev() {
            let jump = nums[i] as usize;
            for j in 1..jump+1 {
                if i+j >= nums.len() {
                    break;
                }
                dp[i] = cmp::min(dp[i], 1+dp[i+j]);
            }
        }
        
        return dp[0];
    }


    // O(n) time,
    // O(1) space,
    // Approach: greedy, 
    pub fn jump(nums: Vec<i32>) -> i32 {
        let mut jumps = 0;
        let mut next_max = 0;
        let mut curr_max = 0;
        
        for i in 0..nums.len() {
            next_max = cmp::max(next_max, i as i32 +nums[i]);
            
            if curr_max >= nums.len()-1 {
                return jumps;
            }
            
            if i == curr_max {
                jumps += 1;
                curr_max = next_max as usize;
            }
        }
        
        return -1;
    }
}